tests/test_top_catch_all.cpp

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (C) 2020 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "mock_logger.h"
+
+#include <multipass/logging/log.h>
+#include <multipass/top_catch_all.h>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <functional>
+#include <memory>
+#include <stdexcept>
+
+namespace mp = multipass;
+namespace mpl = multipass::logging;
+namespace mpt = multipass::test;
+using namespace testing;
+
+namespace
+{
+struct TopCatchAll : public Test
+{
+    void SetUp() override
+    {
+        mock_logger = std::make_shared<StrictMock<mpt::MockLogger>>();
+        mpl::set_logger(mock_logger);
+    }
+
+    void TearDown() override
+    {
+        mock_logger.reset();
+        mpl::set_logger(mock_logger);
+    }
+
+    template <typename Matcher>
+    auto make_cstring_matcher(const Matcher& matcher)
+    {
+        return Property(&mpl::CString::c_str, matcher);
+    }
+
+    auto make_category_matcher()
+    {
+        return make_cstring_matcher(StrEq(category.c_str()));
+    }
+
+    const std::string category = "testing";
+    std::shared_ptr<StrictMock<mpt::MockLogger>> mock_logger = nullptr;
+};
+} // namespace
+
+TEST_F(TopCatchAll, calls_function_with_no_args)
+{
+    int ret = 123, got = 0;
+    EXPECT_NO_THROW(got = mp::top_catch_all("", [ret] { return ret; }););
+    EXPECT_EQ(got, ret);
+}
+
+TEST_F(TopCatchAll, calls_function_with_args)
+{
+    int a = 5, b = 7, got = 0;
+    EXPECT_NO_THROW(got = mp::top_catch_all("", std::plus<int>{}, a, b););
+    EXPECT_EQ(got, a + b);
+}
+
+TEST_F(TopCatchAll, handles_unknown_error)
```

Comment:
Besides `std::runtime_error` and unknown exceptions, I'd suggest to test also for other less common exceptions, such as those raised by ssh. Again, take this as a suggestion.

Reply:
Yup, ok I will. Do you have any specific exception type you could suggest? :)

Reply:
@luis4a0 I suppose you were referring to our own SSH exceptions (libssh itself is C), which are themselves derived from `std::runtime_error`. To account for other `std::exceptions`, I just created a test exception locally. Does that agree with what you had in mind?

Reply:
That seems correct to me, thanks